distributed/worker.py

```diff
@@ -1884,73 +1863,70 @@ def send_task_state_to_scheduler(self, key):
                 "status": "error",
-                "key": key,
-                "thread": self.threads.get(key),
-                "exception": self.exceptions[key],
-                "traceback": self.tracebacks[key],
+                "key": ts.key,
+                "thread": self.threads.get(ts.key),
+                "exception": ts.exception,
+                "traceback": ts.traceback,
             }
         else:
-            logger.error(
-                "Key not ready to send to worker, %s: %s", key, self.task_state[key]
-            )
+            logger.error("Key not ready to send to worker, %s: %s", ts.key, ts.state)
             return
 
-        if key in self.startstops:
-            d["startstops"] = self.startstops[key]
+        if ts.startstops:
+            d["startstops"] = ts.startstops
         self.batched_stream.send(d)
 
-    def put_key_in_memory(self, key, value, transition=True):
-        if key in self.data:
+    def put_key_in_memory(self, ts, value, transition=True):
+        if ts.key in self.data:
+            ts.state = "memory"
             return
 
-        if key in self.actors:
-            self.actors[key] = value
+        if ts.key in self.actors:
+            self.actors[ts.key] = value
 
         else:
             start = time()
-            self.data[key] = value
+            self.data[ts.key] = value
+            ts.state = "memory"
             stop = time()
             if stop - start > 0.020:
-                self.startstops[key].append(
+                ts.startstops.append(
                     {"action": "disk-write", "start": start, "stop": stop}
                 )
 
-        if key not in self.nbytes:
-            self.nbytes[key] = sizeof(value)
+        if ts.key not in self.nbytes:
+            self.nbytes[ts.key] = sizeof(value)
 
-        self.types[key] = type(value)
+        ts.type_ = type(value)
 
-        for dep in self.dependents.get(key, ()):
-            if dep in self.waiting_for_data:
-                if key in self.waiting_for_data[dep]:
-                    self.waiting_for_data[dep].remove(key)
-                if not self.waiting_for_data[dep]:
-                    self.transition(dep, "ready")
+        for dep in ts.dependents:
+            dep.waiting_for_data.discard(ts.key)
+            self.waiting_for_data -= 1
+            if not dep.waiting_for_data:
+                self.transition(dep, "ready")
 
-        if transition and key in self.task_state:
-            self.transition(key, "memory")
-
-        self.log.append((key, "put-in-memory"))
+        self.log.append((ts.key, "put-in-memory"))
 
     def select_keys_for_gather(self, worker, dep):
+        assert isinstance(dep, str)
         deps = {dep}
 
         total_bytes = self.nbytes[dep]
         L = self.pending_data_per_worker[worker]
 
         while L:
             d = L.popleft()
-            if self.dep_state.get(d) != "waiting":
+            if getattr(self.tasks.get(d), "state", None) != "waiting":
```

Comment:
Do we still need this `getattr` call or can we assume this?

Oh, or is this for if `d not in self.tasks`?

Reply:
> Oh, or is this for if `d not in self.tasks`?

Yeah, I should add a comment about that

Reply:
Or just make it more explicit in the first place